pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,399 @@
+}
+
+func decodeMeta(r io.Reader, name string) (metadata.Meta, error) {
+	dec := json.NewDecoder(r)
+	var meta metadata.Meta
+	if err := dec.Decode(&meta); err != nil {
+		return meta, errors.Wrap(err, fmt.Sprintf("failed decoding %s", name))
+	}
+
+	return meta, nil
+}
+
+func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, r *http.Request,
+	logger log.Logger, userBkt objstore.Bucket, tenantID string, blockID ulid.ULID) error {
+	level.Debug(logger).Log("msg", "received request to complete block upload", "content_length", r.ContentLength)
+
+	tmpMetaPath := path.Join(blockID.String(), tmpMetaFilename)
+	rdr, err := userBkt.Get(ctx, tmpMetaPath)
+	if err != nil {
+		if userBkt.IsObjNotFoundErr(err) {
+			return httpError{
+				message:    fmt.Sprintf("upload of block %s not started yet", blockID),
+				statusCode: http.StatusNotFound,
+			}
+		}
+		return errors.Wrap(err, fmt.Sprintf("failed to download %s from object storage", tmpMetaFilename))
+	}
+	defer func() {
+		_ = rdr.Close()
+	}()
+
+	meta, err := decodeMeta(rdr, tmpMetaFilename)
+	if err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
+
+	// Upload meta file so block is considered complete
+	if err := c.uploadMeta(ctx, logger, meta, blockID, block.MetaFilename, userBkt); err != nil {
+		return err
+	}
+
+	if err := userBkt.Delete(ctx, path.Join(blockID.String(), tmpMetaFilename)); err != nil {
+		level.Warn(logger).Log("msg", fmt.Sprintf(
+			"failed to delete %s from block in object storage", tmpMetaFilename), "err", err)
+		return nil
+	}
+
+	level.Debug(logger).Log("msg", "successfully completed block upload")
+	return nil
+}
+
+// sanitizeMeta sanitizes and validates a metadata.Meta object. If a validation error occurs, an error
+// message gets returned, otherwise an empty string.
+func (c *MultitenantCompactor) sanitizeMeta(logger log.Logger, tenantID string, blockID ulid.ULID,
+	meta *metadata.Meta) string {
+	meta.ULID = blockID
+
+	for l, v := range meta.Thanos.Labels {
+		switch l {
+		// Preserve these labels
+		case mimir_tsdb.CompactorShardIDExternalLabel:
+		// Remove unused labels
+		case mimir_tsdb.DeprecatedTenantIDExternalLabel, mimir_tsdb.DeprecatedIngesterIDExternalLabel, mimir_tsdb.DeprecatedShardIDExternalLabel:
+			level.Debug(logger).Log("msg", fmt.Sprintf("removing unused external label from %s", block.MetaFilename),
+				"label", l, "value", v)
+			delete(meta.Thanos.Labels, l)
+		default:
+			level.Warn(logger).Log("msg", fmt.Sprintf("rejecting unsupported external label in %s", block.MetaFilename),
+				"label", l)
+			return fmt.Sprintf("unsupported external label in %s: %s", block.MetaFilename, l)
+		}
+	}
+
+	for _, f := range meta.Thanos.Files {
+		if f.RelPath == block.MetaFilename {
+			continue
+		}
+
+		if !rePath.MatchString(f.RelPath) {
+			level.Warn(logger).Log("msg", fmt.Sprintf("rejecting file with invalid path in %s", block.MetaFilename),
```

Comment:
No need to log this, since we log all httpError messages in `HandleBlockUpload`. (Same is true for all other messages that we return from this function)

Reply:
Done.